index.js

```diff
@@ -1,8 +1,56 @@
+import { getIndexProviderPeerIdFromSmartContract } from './lib/smart-contract-client.js'
+import { getIndexProviderPeerIdFromFilecoinMinerInfo } from './lib/filecoin-rpc-client.js'
+import { rpc } from './lib/rpc-client.js'
 export { MINER_TO_PEERID_CONTRACT_ADDRESS, MINER_TO_PEERID_CONTRACT_ABI } from './lib/constants.js'
+
 /**
  * @param {string} minerId A miner actor id, e.g. `f0142637`
+ * @param {import('ethers').Contract} smartContract The smart contract instance
+ * @param {object} options
+ * @param {number} [options.maxAttempts]
+ * @param {string} [options.rpcUrl]
+ * @param {string} [options.rpcAuth]
+ * @param {(method:string,params:unknown[])=> Promise<unknown> } [options.rpcFn] The RPC function to use
  * @returns {Promise<string>} Miner's PeerId, e.g. `12D3KooWMsPmAA65yHAHgbxgh7CPkEctJHZMeM3rAvoW8CZKxtpG`
  */
-export async function getIndexProviderPeerId(minerId) {
-  return Promise.reject(new Error('Not implemented'))
+export async function getIndexProviderPeerId(
+  minerId,
+  smartContract,
+  {
+    maxAttempts = 5,
+    rpcUrl = 'https://api.node.glif.io/',
+    rpcAuth,
+    rpcFn = async (method, params) => {
+      return await rpc(method, params, { rpcUrl, rpcAuth })
+    },
+  } = {},
+) {
+  try {
+    // Make a concurrent request to both sources: FilecoinMinerInfo and smart contract
+    const [minerInfoResult, contractResult] = await Promise.all([
+      getIndexProviderPeerIdFromFilecoinMinerInfo(minerId, rpcFn, { maxAttempts }),
+      getIndexProviderPeerIdFromSmartContract(minerId, smartContract),
+    ])
```

Comment:
```suggestion
    const [contractResult , minerInfoResult] = await Promise.all([
      getIndexProviderPeerIdFromSmartContract(minerId, smartContract),
      getIndexProviderPeerIdFromFilecoinMinerInfo(minerId, rpcFn, { maxAttempts }),
    ])
```

This is more intuitive, as we prefer the result from the smart contract